sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala

```diff
@@ -522,13 +517,10 @@ case class InputAdapter(child: SparkPlan, isChildColumnar: Boolean)
     child.executeColumnar()
   }
 
-  override def inputRDD: RDD[InternalRow] = {
-    if (isChildColumnar) {
-      child.executeColumnar().asInstanceOf[RDD[InternalRow]] // Hack because of type erasure
-    } else {
-      child.execute()
-    }
-  }
+  // `InputAdapter` can only generate code to process the rows from its child. If the child produces
+  // columnar batches, there must be a `ColumnarToRowExec` above `InputAdapter` to handle it by
+  // overriding `inputRDD`.
```

Comment:
I had misread/misunderstood this originally.

 we are overriding "inputRDDs"  (note the s) in ColumnarToRowExec which is bypassing calling inputRDD here and calling InputAdapter.executeColumnar from ColumnarToRowExec

Reply:
So the `ColumnarToRowExec` doubles as in `InputAdapter`? Wouldn't it be less confusing to replace the `InputAdapter` with `ColumnarToRowExec`?

Reply:
I tried the idea of `replace the InputAdapter with ColumnarToRowExec` before. But then it's a little weird to see `ColumnarToRowExec` act as both the boundary of columnar execution stage and codegen stage. We need to
1. handle `ColumnarToRowExec` specially when planning whole-stage-codegen
2. implement `ColumnarToRowExec.treeString`, which should print this columnar node and remove the whole-stage-codegen mark from its child's treeString

I feel it's simpler to always let `InputAdapter` be the boundary of codegen stage.